File: python/array_with_elements_not_equal_to_average_of_neighbors.py
class Solution:
    def rearrangeArray2(self, nums: list[int]) -> list[int]:
        for i in range(1, len(nums) - 1):
            if 2 * nums[i] == nums[i - 1] + nums[i + 1]:
                nums[i], nums[i + 1] = nums[i + 1], nums[i]

        for i in reversed(range(1, len(nums) - 1)):
            if 2 * nums[i] == nums[i - 1] + nums[i + 1]:
                nums[i], nums[i - 1] = nums[i - 1], nums[i]

        return nums

    def rearrangeArray(self, nums: list[int]) -> list[int]:
        result: list[int] = list()
        nums.sort()
        l: int = 0
        r: int = len(nums) - 1
        while len(result) != len(nums):
            result.append(nums[l])
            l += 1

            if l < r:
                result.append(nums[r])
                r -= 1

        return result


def test_rearrangeArray_case_1():
    # arrange
    nums: list[int] = [1, 2, 3, 4, 5]
    expected: list[int] = [1, 5, 2, 4, 3]

    # act
    solution = Solution()
    actual = solution.rearrangeArray(nums)

    # assert
    assert expected == actual


def test_rearrangeArray_case_2():
    # arrange
    nums: list[int] = [6, 2, 0, 9, 7]
    expected: list[int] = [0, 9, 2, 7, 6]

    # act
    solution = Solution()
    actual = solution.rearrangeArray(nums)

    # assert
    assert expected == actual


def test_rearrangeArray_case_3():
    # arrange
    nums: list[int] = [10, 7, 5, 4, 3]
    expected: list[int] = [10, 5, 7, 3, 4]

    # act
    solution = Solution()
    actual = solution.rearrangeArray2(nums)

    # assert
    assert expected == actual

File: rust/src/code/array_with_elements_not_equal_to_average_of_neighbors.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity: O(nlogn)
    pub fn rearrange_array(mut nums: Vec<i32>) -> Vec<i32> {
        nums.sort();
        let mut result: Vec<i32> = vec![0; nums.len()];
        let h: usize = if nums.len() % 2 == 0 {
            nums.len() / 2
        } else {
            (nums.len() + 1) / 2
        };
        let mut l: usize = 0;
        let mut r: usize = h;
        for i in 0..result.len() {
            if i % 2 == 0 && l < h {
                result[i] = nums[l];
                l += 1;
            } else if i % 2 > 0 && r < nums.len() {
                result[i] = nums[r];
                r += 1;
            }
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_rearrange_array_case_1() {
        // arrange
        let nums: Vec<i32> = vec![1, 2, 3, 4, 5];
        let expected: Vec<i32> = vec![1, 5, 2, 4, 3];

        // act
        let actual = Solution::rearrange_array(nums);

        // assert
        assert_eq!(actual, expected);
    }
}
